runtime/src/mem/jemalloc/mem-jemalloc.c

```diff
@@ -70,46 +72,79 @@ static inline void* alignHelper(void* base_ptr, size_t offset, size_t alignment)
+    pthread_mutex_lock(&heap.alloc_lock);
+
+    // compute our current aligned pointer into the shared heap
+    //
+    //   jemalloc 4.5.0 man: "The alignment parameter is always a power of two
+    //   at least as large as the chunk size."
+    cur_chunk_base = alignHelper(heap.base, heap.cur_offset, alignment);
+
+    // jemalloc 4.5.0 man: "If chunk is not NULL, the returned pointer must be
+    // chunk on success or NULL on error"
+    if (chunk && chunk != cur_chunk_base) {
+      pthread_mutex_unlock(&heap.alloc_lock);
+      return NULL;
+    }
 
-  // compute our current aligned pointer into the shared heap
-  //
-  //   jemalloc 4.5.0 man: "The alignment parameter is always a power of two at
-  //   least as large as the chunk size."
-  cur_chunk_base = alignHelper(heap.base, heap.cur_offset, alignment);
+    cur_heap_size = (uintptr_t)cur_chunk_base - (uintptr_t)heap.base;
 
-  // jemalloc 4.5.0 man: "If chunk is not NULL, the returned pointer must be
-  // chunk on success or NULL on error"
-  if (chunk && chunk != cur_chunk_base) {
-    pthread_mutex_unlock(&heap.alloc_lock);
-    return NULL;
-  }
+    // If there's not enough space on the heap for this allocation, return NULL
+    if (size > heap.size - cur_heap_size) {
+      pthread_mutex_unlock(&heap.alloc_lock);
+      return NULL;
+    }
 
-  cur_heap_size = (uintptr_t)cur_chunk_base - (uintptr_t)heap.base;
+    // Update the current pointer, now that we've past any early returns.
+    heap.cur_offset = cur_heap_size + size;
 
-  // If there's not enough space on the heap for this allocation, return NULL
-  if (size > heap.size - cur_heap_size) {
+    // now that cur_heap_offset is updated, we can unlock
     pthread_mutex_unlock(&heap.alloc_lock);
-    return NULL;
-  }
+  } else {
+    //
+    // Get a dynamic extension chunk.
+    //
+    cur_chunk_base = chpl_comm_regMemAlloc(size);
 
-  // Update the current pointer, now that we've past any early returns.
-  heap.cur_offset = cur_heap_size + size;
+    if (cur_chunk_base == NULL) {
+      return NULL;
+    }
+
+    //
+    // Localize the new memory via first-touch, by storing to each page.
+    // This will give the memory affinity to the NUMA domain (if any)
+    // we were running on when we allocated it.  It also commits the
+    // memory, in the jemalloc sense.
+    //
+    for (int i = 0; i < size; i += heap.page_size) {
+      ((char*) cur_chunk_base)[i] = 0;
+    }
 
-  // now that cur_heap_offset is updated, we can unlock
-  pthread_mutex_unlock(&heap.alloc_lock);
+    chpl_comm_regMemPostAlloc(cur_chunk_base, size);
+
+    // Support useUpMemNotInHeap by hiding the fixed/dynamic distinction
+    heap.base = cur_chunk_base;
+    heap.size = size;
+    heap.cur_offset = 0;
+  }
 
   // jemalloc 4.5.0 man: "Zeroing is mandatory if *zero is true upon entry."
   if (*zero) {
-     memset(cur_chunk_base, 0, size);
+    memset(cur_chunk_base, 0, size);
```

Comment:
You can skip this in the case where you've already zero'd above, right?

Reply:
Not necessarily.  If the memory delivered by the comm layer a few lines previously is on old pages instead of new ones, then it may have non-0 contents and need zeroing.  However, we could look into whether there's a way to tell this and avoid the zeroing cost if the pages are new, because new pages are guaranteed to be zero I believe (though I'd have to go search for confirmation of that).  Agreed it would be nice to avoid the zeroing code here if we can, because these are big chunks.

Reply:
Nah, I missed that the above code wasn't zeroing the entire chunk, so ignore this comment